New guide page for Selenium migrations (#5780)
cypress-io/cypress-documentation: docs/guides/end-to-end-testing/migration/selenium-to-cypress.mdx

reporting with access to screenshots, videos, and time travel debugging.

Furthermore, the addition of [Component Testing](/guides/component-testing/overview) to Cypress has fostered
[collaboration](https://www.cypress.io/blog/2023/01/26/what-cypress-component-testing-means-for-qa-teams) between testing and development teams,
enabling them to work with the same tools and language. This alignment streamlines the testing process and promotes
consistency and efficiency across teams.

## Top 5 reasons to migrate

While migration is often perceived as a hassle, investing in Cypress is undoubtedly worthwhile.
In fact, migrating to Cypress may not even pose a significant challenge.

### 1. Reduction of Flakiness

One of the primary benefits of Cypress is its ability to reduce flakiness in tests. Occasional
test failures can undermine trust in the testing framework. Cypress addresses this concern by
offering features such as automatic retries and implicit waiting on the application, effectively
minimizing flakiness and enhancing test reliability.

### 2. Ease of Setup

Unlike traditional test automation frameworks that require cumbersome setup processes, Cypress

> streamlines the setup with just two Node.js commands: ['npm install'](/guides/getting-started/installing-cypress#npm-install)

Reviewer: ```suggestion
streamlines the setup with just two Node.js commands: ['npm install'](/guides/getting-started/installing-cypress#Installing)
```

Make the link more generic to the section about Installing.

community, allow users to tailor their testing framework to specific project requirements, ensuring flexibility and scalability.

### 4. Strong Community Support

Cypress boasts a vibrant community that contributes plugins and prioritizes documentation.
With extensive documentation, support initiatives like the [Ambassador program](https://www.cypress.io/ambassadors), and a [Discord
community](https://discord.com/invite/cypress) of over 10,000 testers and developers, Cypress users benefit from wealth of resources and
assistance, making it a highly supported testing framework.

### 5. Ease of Use

Compared to Selenium, Cypress offers a more intuitive and user-friendly experience. Features like the [selector
playground](/guides/core-concepts/cypress-app#Selector-Playground) and
[built-in screenshots](/guides/guides/screenshots-and-videos) simplify test writing and
debugging, allowing users to focus on creating effective tests without the need for constant navigation between the test server and browser. This
streamlined approach enhances productivity and accelerates the testing process. Cypress [Test Replay](/guides/cloud/test-replay), which provides

> the ability to time travel and interactively debug tests, is not a capability that Selenium provides. While debugging is never the most enjoyable

Reviewer 2: It's worth noting the context here because time travel debugging is a core part of the test runner itself (the local experience, etc)

```suggestion
the ability to time travel and interactively debug tests in your CI/CD pipeline, is not a capability that Selenium provides. While debugging is never the most enjoyable
```

in Cypress. This functionality is [built in](/guides/core-concepts/retry-ability).

The next vital part to migrate would be the [assertions](/guides/references/assertions). We can all agree assertions
are moderately vital to a test automation framework. The next piece of code is a
python Selenium assertion to validate if a button is visible. While it is possible
to add third-party libraries, the built-in assertions of Cypress can tackle a wide
range of situations.

```python
assert driver.find_element(By.ID, "myButton").is_displayed()
```

In Cypress, the same command would be:

```javascript
cy.get('#myButton').should('be.visible')
```

Considering the easy and clear syntax of Cypress, transferring all the selectors
might take the longest. And while copying and pasting the selectors is definitely
an option, rewriting the selectors might be required to conform to the best practices.

##### iframe

It's acknowledged that Cypress could handle iframes more effectively _out-of-the-box_. But, with the support of a robust
community, solutions are readily available. In the case of iframes, the [cypress-iframe](https://www.npmjs.com/package/cypress-iframe)

> plugin was developed. This plugin will add a custom command to find and interact with an iframe.

Reviewer: - PR https://github.com/cypress-io/cypress-documentation/pull/5789 from @dingraham
 proposes to deprecate the outdated [cypress-iframe](https://www.npmjs.com/package/cypress-iframe) plugin.

##### File upload

Another community-based solution worth mentioning is file uploads. Although Cypress offers a built-in solution for
handling [file uploads](/api/commands/selectfile), it may not be compatible with every upload component available.
While `cy.get('input[type=file]').selectFile('file.json')` suffices in most scenarios, every so often, the web application requires a different

> approach. In such instances, plugins like this prove invaluable: [cypress-file-upload](https://www.npmjs.com/package/cypress-file-upload).

Reviewer: [cypress-file-upload](https://www.npmjs.com/package/cypress-file-upload) is not listed on https://docs.cypress.io/plugins

It was last updated in June 2021 and is configured for legacy Cypress 3.0.0. The documentation is at legacy level.
Does it work with Cypress `13.x`?

Author: you're correct.  I found this blog which advises we replace the plugin with the in-built selectFile command.
https://www.cypress.io/blog/2022/01/19/uploading-files-with-selectfile

Reviewer: Good find! Your blog article also links back to the documentation site https://docs.cypress.io/guides/references/migration-guide#Migrating-from-cypress-file-upload-to-selectFile as part of the section https://docs.cypress.io/guides/references/migration-guide#Migrating-to-Cypress-100

are moderately vital to a test automation framework. The next piece of code is a
python Selenium assertion to validate if a button is visible. While it is possible
to add third-party libraries, the built-in assertions of Cypress can tackle a wide
range of situations.

```python
assert driver.find_element(By.ID, "myButton").is_displayed()
```

In Cypress, the same command would be:

```javascript
cy.get('#myButton').should('be.visible')
```

Considering the easy and clear syntax of Cypress, transferring all the selectors
might take the longest. And while copying and pasting the selectors is definitely
an option, rewriting the selectors might be required to conform to the best practices.

##### Multiple tabs

When you have the need to interact with multiple browser tabs in Cypress, an official plugin has been provided to accomplish this
goal. The [Puppeteer plugin](https://github.com/cypress-io/cypress/tree/develop/npm/puppeteer) utilizes Puppeteer's browser API
with one command within Cypress.

##### File upload

Because Selenium cannot interact with the file upload dialog, it provides a way to upload files without opening the dialog. Cypress

> provides support uploading files through the built-in ['selectFile'](/api/commands/selectfile) command. There are

Reviewer 2: We try to always refer to the commands using their syntax for uniformity. 

```suggestion
provides support uploading files through the built-in [`.selectFile()`](/api/commands/selectfile) command. There are
```

## Integration with CI/CD Pipeline

### Integration with the current Pipeline

Cypress can be integrated into any pipeline technology using its headless runner. This headless runner leverages
the Electron browser, ensuring minimal adaptation of the current CI/CD. Given the widespread usage of Docker in
most pipelines today, the Cypress pipeline is no exception. Integration becomes effortless by employing a Node.js
docker image or one provided by Cypress via Docker Hub. These Docker images, not only furnished by Cypress,
include various versions of Chrome, Firefox, and Edge.

Executing the following command initiates a headless run against the Electron browser. Upon completion, a succinct
table is generated, detailing the number of tests for each 'cy.js' file, in addition to the failed, skipped and passed
tests with the run time of each specific 'cy.js' file. At the bottom of the table, a comprehensive summary displays the
total number of tests run, the duration of the run, and the counts of failed, passed, and skipped tests.

```bash
npx cypress run
```

This would completely replace the Selenium step in your current CI/CD workflow.

> Learn more about running Cypress tests in Continuous Integration [here](/guides/continuous-integration/introduction).

Reviewer 2: GTM can chime in here but I believe this allows for better SEO when crawling these pages.

```suggestion
Learn more about [running Cypress tests in Continuous Integration](/guides/continuous-integration/introduction).
```

documentation and troubleshooting guides. Supported by an active community, Selenium benefits
from robust documentation across multiple languages including Java, JavaScript, Python, C#, and
Ruby. Its widespread adoption is further evidenced by its integration into various low-code
tools, underscoring its versatility in browser interaction.

A key differentiator of Cypress, not only from Selenium but also from other test automation frameworks,
is its [unique approach](/guides/overview/key-differences#Architecture) of running tests
inside an interactive browser rather than against one. This approach [minimizes flakiness](/guides/overview/key-differences#Flake-resistant)
and grants users greater control over both the framework and the application under test. Cypress's ability to wait for page loading without explicit
declaration allows for patient execution until the application is fully ready. Additionally, Cypress simplifies test recording
without requiring complex configurations or setups, facilitating easy troubleshooting and bug
reporting with access to screenshots, videos, and time travel debugging.

> Furthermore, the addition of [component testing](/guides/component-testing/overview) to Cypress has fostered

Author: You're right.  It's proper noun elsewhere, so incorporating this change immediately

[collaboration](https://www.cypress.io/blog/2023/01/26/what-cypress-component-testing-means-for-qa-teams) between testing and development teams,
enabling them to work with the same tools and language. This alignment streamlines the testing process and promotes
consistency and efficiency across teams.

> ## Top 5 reasons to migrate

Reviewer 2: Maybe enumerating these helps drive it home? Generally looking for ways for this content to be more easily parsed while still useful.

While migration is often perceived as a hassle, investing in Cypress is undoubtedly worthwhile.
In fact, migrating to Cypress may not even pose a significant challenge.

### Reduction of Flakiness

One of the primary benefits of Cypress is its ability to reduce flakiness in tests. Occasional
test failures can undermine trust in the testing framework. Cypress addresses this concern by
offering features such as automatic retries and implicit waiting on the application, effectively

> minimizing flakiness and enhancing test reliability.

Author: I saved this to spawn off on internal discussion with the app team this week.  If we make adjustments we'll do it as a second pass after we deploy the initial material

### Ease of Setup

Unlike traditional test automation frameworks that require cumbersome setup processes, Cypress
streamlines the setup with just two Node.js commands: ['npm install'](/guides/getting-started/installing-cypress#npm-install)
and ['npx cypress open'](/guides/getting-started/opening-the-app#cypress-open).
Within seconds, users can have a fully functional framework up and running. Since Cypress utilizes
locally installed browsers, there's no need for complex configurations or additional installations.

### Infinitely Expandable

Recognizing the diverse needs of web applications, Cypress offers extensibility through various Node.js

> [plugins](/plugins#__docusaurus_skipToContent_fallback). These plugins, many of which are contributed by the

Reviewer: ```suggestion
[plugins](/plugins). These plugins, many of which are contributed by the
```

community, allow users to tailor their testing framework to specific project requirements, ensuring flexibility and scalability.

### Strong Community Support

Cypress boasts a vibrant community that contributes plugins and prioritizes documentation.
With extensive documentation, support initiatives like the [Ambassador program](https://www.cypress.io/ambassadors), and a [Discord
community](https://discord.com/invite/cypress) of over 10,000 testers and developers, Cypress users benefit from wealth of resources and
assistance, making it a highly supported testing framework.

### Ease of Use

> Compared to Selenium, Cypress offers a more intuitive and user-friendly experience. Features like the [selector

Author: Agree with you on Test Replay.  I'm making that change now and will let marketing know afterwards.  Selenium does have Selenium IDE which offers similar capabilities to Cypress Studio, so leaving that out for now

> Compared to Selenium, Cypress offers a more intuitive and user-friendly experience. Features like the [selector

Reviewer 2: Got it. Was thinking less about apples to apples but more about a quick way to get up and running with Cypress test coverage/API that could assist in the migration. But could be a stretch.

playground](/guides/core-concepts/cypress-app#Selector-Playground) and
[built-in screenshots](/guides/guides/screenshots-and-videos#__docusaurus_skipToContent_fallback) simplify test writing and
debugging, allowing users to focus on creating effective tests without the need for constant navigation between the test server and browser. This
streamlined approach enhances productivity and accelerates the testing process. While debugging is never the most enjoyable task in
development, Cypress is here to assist us. With its clear and concise error messages and multitude of plugins for feedback,
debugging has never been easier.

In summary, migrating to Cypress offers a host of benefits, including reduced flakiness, simplified setup,
extensibility, strong community support, and enhanced ease of use. With Cypress, testing becomes not only more
reliable and efficient but also more enjoyable for developers and QA professionals alike.

## Evaluating Suitability for Migration

> In order for a migration to proceed smoothly, it is crucial to conduct a thorough analysis of the existing test cases

Author: Changed it from 'crucial' to 'important' in my local copy.  

Regarding all or nothing, the next couple of sentences dive right into that concept.  I did change the next sentence to be more explicit about that advantage:
"In order for a migration to proceed smoothly, it is important to conduct a thorough analysis of the existing test cases and framework functionality. This analysis serves as a cornerstone for determining the priority of test case migration since not all test cases need to be migrated at once"

This incremental approach facilitates faster migration of additional tests and reduces the complexity of larger, more intricate tests.  
Alternatively, if the current framework does not utilize the Page Object Model, selector migration will occur naturally as part of
migrating the actual codebase.

It's important to note that if the Selenium framework relies on XPath, migration will necessitate changing those selectors.
While the 'cypress-xpath' plugin remains functional, it is deprecated and may not be the optimal solution in the long term.

Alternatively, this could present an opportune moment to adopt the Page Object Model methodology,
streamlining test maintenance and enhancing readability.

## Getting started with Cypress

Installing and configuring Cypress is very easy and we have provided step-by-step guides to help you get started.

You will first [install Cypress](/guides/getting-started/installing-cypress).
You will then [open the Cypress application](/guides/getting-started/opening-the-app).

Check out our in-depth [E2E getting started guide](/guides/end-to-end-testing/writing-your-first-end-to-end-test).

And now you're all set to start the migration.

> ## Strategy and Considerations

Author: Pulling this out into the separate discussion, also.  I agree this might be the order we want to use in a second deployment

backend dependencies and enhance framework stability.

In larger applications, locating specific tests within the framework can prove challenging. Therefore, establishing
guidelines for test organization is essential for creating a maintainable framework. Some teams prefer organizing tests
based on application functionality, while others opt for organization by webpage. While performance and functionality
remain unaffected, clear guidelines on test placement are imperative for framework maintainability.

These considerations serve as a roadmap toward crafting a well-structured, easily maintainable, and enjoyable-to-use framework.
Investing time in contemplation and extensive team discussions regarding these questions will yield significant dividends in the long run.

With thorough preparation underway, it's time to initiate the actual migration process.

One of the first things to migrate would be all page-objects or selectors, and depending on the frontend framework the
developers are using, this could be mostly copy and paste. For instance, in Vue.js the selector in Selenium would be

> `vl-checkbox` while in Cypress this would be `.vl-checkbox`. Granted, the difference is annoying to fix,

Reviewer 2: Still curious about this one. It almost implies the use of css selectors which we advise against as a best practice. It might make sense just as a simple reference example, I just wasn't clear on the origin.

While Selenium will run a class or different scenarios, Cypress will run every `it` in a `*.cy.js` file.
So, the flow of a test should be transferred inside an `it`. To make matters more structured, it is possible to add
several `it`s inside a `describe`. This allows you as the user to create a subdivision akin to test suites and test cases.

```javascript
describe('General information', () => {
  beforeEach(() => {
    cy.get('#loginButton').click()
  })

  it('Header is visible', () => {
    cy.get('#header').should('be.visible')
  })

  it('Footer is visible', () => {
    cy.get('#footer').should('be.visible')
  })
})
```

<br />
Another big change is that you no longer need any WebDriver definitions. Cypress
takes care of opening and closing the browser without any explicit prompting, unlike
Selenium. Any code that is used to define the WebDriver is now obsolete.

```java
 WebDriver driver = new ChromeDriver();

 WebDriverWait wait = new WebDriverWait(driver, 10);

 driver.quit();
```

<br />
Next would be to replace all the potential actions. The following code is fairly
standard in Selenium frameworks.

> 

Author: pulling this out to a separate convo

```python
wait.until(EC.element_to_be_clickable((By.ID, "myButton"))).click()
```

<br />
It is fairly easy to change this line of code to work inside a Cypress framework.
It could even be done with a find and replace. In Cypress, the command would be the
following:

```Javascript
cy.get('#myButton').click()
```

<br />
Any wait until functionality that you have been using in Selenium is no longer required
in Cypress. This functionality is built in.

> The next vital part to migrate would be the assertions. We can all agree assertions

Author: Added.  Also put in a link for retry-ability

are moderately vital to a test automation framework. The next piece of code is a
python Selenium assertion to validate if a button is visible. While it is possible
to add third-party libraries, the built-in assertions of Cypress can tackle a wide
range of situations.

```python
assert driver.find_element(By.ID, "myButton").is_displayed()
```

In Cypress, the same command would be:

```javascript
cy.get('#myButton').should('be.visible')
```

Considering the easy and clear syntax of Cypress, transferring all the selectors
might take the longest. And while copying and pasting the selectors is definitely
an option, rewriting the selectors might be required to conform to the best practices.

##### Plugins

One significant advantage of implementing Cypress is gaining access to its

> [extensive library of plugins](/plugins#__docusaurus_skipToContent_fallback).

Reviewer: ```suggestion
[extensive library of plugins](/plugins).
```

This collection of community-built solutions can effectively address virtually any issue you might encounter.

##### Iframe

It's widely acknowledged that Cypress doesn't handle iframes effectively. However, with the support of a robust
community, solutions are readily available. In the case of iframes, the [cypress-iframe - npm](https://www.npmjs.com/package/cypress-iframe)
plugin was developed. This plugin will add a custom command to find and interact with an iframe.

##### File upload

Another community-based solution worth mentioning is file uploads. Although Cypress offers a built-in solution for

> handling file uploads, it may not be compatible with every upload component available. While `cy.get().selectFile()`

Author: incorporated

suffices in most scenarios, every so often, the web application requires a different approach. In such instances,

> plugins like this prove invaluable: [cypress-file-upload - npm](https://www.npmjs.com/package/cypress-file-upload).

Author: unclear, but I've removed it

plugins. These plugins, many of which are contributed by the community, allow users to tailor their testing
framework to specific project requirements, ensuring flexibility and scalability.

### Strong Community Support

Cypress boasts a vibrant community that not only contributes plugins but also prioritizes documentation.
With extensive documentation and support initiatives like the ambassador program, Cypress users benefit from a
wealth of resources and assistance, making it a highly supported testing framework.

### Ease of Use

Compared to Selenium, Cypress offers a more intuitive and user-friendly experience. Features like the selector
playground and built-in screenshots simplify test writing and debugging, allowing users to focus on creating effective
tests without the need for constant navigation between the test server and browser. This streamlined approach enhances
productivity and accelerates the testing process.

In summary, migrating to Cypress offers a host of benefits, including reduced flakiness, simplified setup,
extensibility, strong community support, and enhanced ease of use. With Cypress, testing becomes not only more

> reliable and efficient but also more enjoyable for developers and QA professionals alike."

Reviewer: ```suggestion
reliable and efficient but also more enjoyable for developers and QA professionals alike.
```

[start a discussion here](https://github.com/cypress-io/cypress/discussions/new?category=general).

:::

Taking a closer look at Selenium WebDriver reveals its longevity and extensive library of
documentation and troubleshooting guides. Supported by an active community, Selenium benefits
from robust documentation across multiple languages including Java, JavaScript, Python, C#, and
Ruby. Its widespread adoption is further evidenced by its integration into various low-code
tools, underscoring its versatility in browser interaction.

A key differentiator of Cypress, not only from Selenium but also from other test automation frameworks,
is its unique approach of running tests inside an interactive browser rather than against one.
This approach minimizes flakiness and grants users greater control over both the framework and the
application under test. Cypress's ability to wait for page loading without explicit declaration allows
for patient execution until the application is fully ready. Additionally, Cypress simplifies test recording
without requiring complex configurations or setups, facilitating easy troubleshooting and bug
reporting with access to screenshots and videos.

> Furthermore, the recent addition of component testing to Cypress has fostered collaboration between testing

Reviewer: ```suggestion
Furthermore, the addition of component testing to Cypress has fostered collaboration between testing
```

Component testing was officially released as GA (Generally Available) in [cypress@11.0.0](https://docs.cypress.io/guides/references/changelog#11-0-0) in Aug 2022 for React, Next.js, Angular, and Vue. I wouldn't use "recent" for an event which happened more than one year ago.

and development teams, enabling them to work with the same tools and language. This alignment not only
streamlines the testing process but also promotes consistency and efficiency across teams.

## Top 5 reasons to migrate

While migration is often perceived as a hassle, investing in Cypress is undoubtedly worthwhile.
In fact, migrating to Cypress may not even pose a significant challenge.

### Reduction of Flakiness

One of the primary benefits of Cypress is its ability to reduce flakiness in tests. Occasional
test failures can undermine trust in the testing framework. Cypress addresses this concern by
offering features such as automatic retries and implicit waiting on the application, effectively
minimizing flakiness and enhancing test reliability.

### Ease of Setup

Unlike traditional test automation frameworks that require cumbersome setup processes, Cypress

> streamlines the setup with just two Node.js commands: 'npm install' and 'npx cypress run'.

Reviewer: ```suggestion
streamlines the setup with just two commands: 'npm install' and 'npx cypress open'.
```
npm is bundled with Node.js, however it is not the same thing. Cypress can also use Yarn and pnpm package managers
`cypress run` is not a setup command, `cypress open` does configuration if the project has not been configured

Within seconds, users can have a fully functional framework up and running. Since Cypress utilizes
locally installed browsers, there's no need for complex configurations or additional installations.

> ### Infinetely Expandable

Reviewer: ```suggestion
### Infinitely Expandable
```

This can be inside the application directory, or you could consider Cypress as a separate entity. For the purpose of this guide,
Cypress will be considered a separate framework, outside the application under test. There is no difference in installation or use.
The reason this guide will handle it as a standalone application is because it makes screenshots clearer.

<DocsImage
  src="/img/guides/migrating-to-cypress/InstallCypress.png"
  alt="install Cypress"
/>

<DocsImage
  src="/img/guides/migrating-to-cypress/OpenCypressApp.png"
  alt="open the Cypress app"
/>

<br />
The first choice will be the testing type used by the framework, which in this migration
guide will be E2E testing. After selecting an option, Cypress will inform you that
it has created several directories and files.

<DocsImage
  src="/img/guides/migrating-to-cypress/WelcomeToCypress.png"
  alt="welcome to Cypress screen"
/>

<br />
<DocsImage
  src="/img/guides/migrating-to-cypress/LookForCypressInstallFiles.png"
  alt="look for Cypress install files and folders"
/>

<br />
<br />
Once all those files have been created, all that is left is selecting a browser

> to work against. After selecting a browser, Cypress will suggest creating an
> example spec, or you can jump right into it.

Reviewer: ```suggestion
to work against. After selecting a browser, Cypress will suggest creating
example specs, or you can jump right into it.
```

The user is offered "Scaffold example specs" (there are 20 of them) or "Create new spec" which is a single simple spec.

<DocsImage
  src="/img/guides/migrating-to-cypress/CreateFirstSpec.png"
  alt="create your first spec"
/>

<br />
<br />

> If you've selected 'create an example spec', Cypress should immediately start

Reviewer: ```suggestion
If you've selected 'Create new spec', Cypress should immediately start
```

See the screenshot

running that spec file.

<DocsImage
  src="/img/guides/migrating-to-cypress/CypressRunSpec.png"
  alt="Cypress run your first spec"
/>

<br />
And now you're all set to start the migration.

## Migrating test cases

> If the current automation framework supports cucumber, move towards the section `Migrating a cucumber framework`.

Reviewer: ```suggestion
If the current automation framework supports cucumber, move towards the section [Migrating test cases with Cucumber](#Migrating-test-cases-with-Cucumber).
```

One of the first things to migrate would be all page-objects or selectors, depending on the frontend framework the
developers are using, this could be mostly copy and paste. For instance, in Vue.js the selector in Selenium would be
`vl-checkbox` while in Cypress this would be `.vl-checkbox`. Granted, the difference is annoying to fix,
but you could migrate all selectors in record time.

One of the most significant changes between Selenium and Cypress is the actual testing syntax.
While Selenium will run a class or different scenarios, Cypress will run every `it` in a `*.cy.js` file.
So, the flow of a test should be transferred inside an `it`. To make matters more structured, it is possible to add
several `it`s inside a `describe`. This allows you as the user to create a subdivision akin to Test suites and test cases.

```javascript
describe('General information', () => {
  beforeEach(() => {
    cy.get('#loginButton').click()
  })

  it('Header is visible', () => {
    cy.get('#header').should('be.visible')
  })

  it('Footer is visible', () => {
    cy.get('#footer').should('be.visible')
  })
})
```

<br />

> Another big change is, you no longer need any WebDriver definitions. Cypress takes

Reviewer: ```suggestion
Another big change is that you no longer need any WebDriver definitions. Cypress takes
```

Although it alters the conventional `describe` and `it` functionality, the adjustment process should be relatively straightforward.

The main difference would be that instead of adding the whole flow into a single `it` section, you add the different steps into
an `it` like section, where the `it` is replaced with the standard Gherkin syntax.

The first snippet comprises plain Cypress code, whereas the second snippet demonstrates the cucumber version

```javascript
it("I should see a search bar", () => {
  cy.visit("https://www.duckduckgo.com");
  cy.get("input").should(
    "have.attr",
    "placeholder",
    "Search the web without being tracked"
  )
```

```javascript
import { When, Then } from '@badeball/cypress-cucumber-preprocessor'

When('I visit duckduckgo.com', () => {
  cy.visit('https://www.duckduckgo.com')
})

Then('I should see a search bar', () => {
  cy.get('input').should(
    'have.attr',
    'placeholder',
    'Search the web without being tracked'
  )
})
```

<br />
As evident, the actual code or logic remains unchanged. The difference lies in the
structure of the spec file and the framework's architecture. [Run cucumber/gherkin-syntaxed

> specs with Cypress](https://github.com/badeball/cypress-cucumber-preprocessor)

Reviewer: ```suggestion
specs with Cypress](https://github.com/badeball/cypress-cucumber-preprocessor).
```

## Integration with CI/CD Pipeline

### Integration with the current Pipeline

Cypress can be integrated into any pipeline technology using its headless runner. This headless runner leverages
the Electron browser, ensuring minimal adaptation of the current CI/CD. Given the widespread usage of Docker in

> most pipelines today, the new Cypress pipeline is no exception. Integration becomes effortless by employing a Node.js

Reviewer: ```suggestion
most pipelines today, the Cypress pipeline is no exception. Integration becomes effortless by employing a Node.js
```

docker image or one provided by Cypress via Docker Hub. These Docker images, not only furnished by Cypress,
include various versions of Chrome, Firefox, and Edge.

Executing the following command initiates a headless run against the Electron browser. Upon completion, a succinct

> table is generated, detailing the amount of tests for each cy.js file, in addition to the failed, skipped and passed
> tests with the run time of that specific cy.js file. At the bottom of the table, a comprehensive summary displays the

Reviewer: ```suggestion
table is generated, detailing the number of tests for each cy.js file, in addition to the failed, skipped and passed
tests with the run time of each specific cy.js file. At the bottom of the table, a comprehensive summary displays the
```

total number of tests run, the duration of the run, and the counts of failed, passed, and skipped tests.

```bash

> Npx cypress run

Reviewer: ```suggestion
npx cypress run
```

Mixed case works on Windows (which is case-insensitive / case-preserving) but not on Linux (case-sensitive). The correct command is `npx`. (Depending on the package manager, yarn or pnpm could also be used here.)

Exactly how to run Cypress in CI/CD depends on the CI provider and there is detailed documentation available for this under https://docs.cypress.io/guides/continuous-integration/introduction.

```

This would completely replace the Selenium step in your current CI/CD workflow.

### Cypress Cloud

The notable distinction in integration between Cypress and Selenium emerges when discussing Cypress Cloud specifically.

Cypress.io offers its tailored pipeline structure, meticulously designed to execute tests efficiently.
Accompanying this structure are several advanced features that warrant attention during
migration to Cypress. While it's feasible to harness all the advantages of migration without Cypress Cloud,
it would be remiss not to explore the potential benefits afforded by its features.

Some of the core features include:

> ##### [Test Replay](https://docs.cypress.io/guides/cloud/test-replay)

Reviewer: ```suggestion
##### [Test Replay](/guides/cloud/test-replay)
```

One of the major difficulties in troubleshooting headless testing is having to reproduce the failure in your local
environment in order to debug failures in CI. Cypress Cloud resolves this issue with Test Replay. Test Replay's
interactive interface allows for faster debugging, enabling your team to time travel through your application under
test to rapidly identify and resolve issues that surfaced in CI. This eliminates the need to reproduce CI issues
locally so you can swiftly get back to shipping valuable software for your users. Gain visibility into the events
leading up to crashes or failures by inspecting the DOM, network events, and console logs of your application under
test exactly as they ran in CI.

> ##### [Parallelization](https://docs.cypress.io/guides/cloud/smart-orchestration/parallelization)

Reviewer: ```suggestion
##### [Parallelization](/guides/cloud/smart-orchestration/parallelization)
```

To enhance testing efficiency further, Cypress Cloud offers a built-in solution for test parallelization. With a simple
toggle, tests can run in parallel, eliminating the need for complex configurations to run tests concurrently.

Cypress will assign each spec file to an available machine based on our

> [balance strategy](https://docs.cypress.io/guides/cloud/smart-orchestration/load-balancing#Balance-strategy).

Reviewer: ```suggestion
[balance strategy](/guides/cloud/smart-orchestration/load-balancing#Balance-strategy).
```

Due to this balance strategy, the run order of the spec files is not guaranteed when parallelized.

> ##### [Analytics and Reporting](https://docs.cypress.io/guides/cloud/analytics)

Reviewer: ```suggestion
##### [Analytics and Reporting](/guides/cloud/analytics)
```

The last step in pipeline integration is reporting the results. While most pipelines can report if there is a failure,
precise numbers require a bit more configuration. One of the most popular plugins to show reports is
mochawesome ([cypress-mochawesome-reporter - npm](https://www.npmjs.com/package/cypress-mochawesome-reporter)). This
plugin can easily be integrated into most pipeline infrastructure, with the correct CI/CD plugin.

But even without the mochawesome package, any pipeline infrastructure can register the error code (error code 1) if
there are test failures. And as soon as this error code is found, webhooks can trigger a notification in your medium of choice.

> Cypress Cloud can also serve as an automated pipeline integration, furnishing [built-in analytics and reporting capabilities]
> (https://docs.cypress.io/guides/cloud/analytics). This feature grants you visibility into metrics such as run duration,

Reviewer: ```suggestion
Cypress Cloud can also serve as an automated pipeline integration, furnishing [built-in analytics and reporting capabilities](/guides/cloud/analytics). This feature grants you visibility into metrics such as run duration,
```

negates the necessity of adding another package or dependency, as Cypress Cloud seamlessly interacts with the tests.

## Conclusion

In this migration guide, we've delved into a diverse array of topics. Beginning with a brief overview of the advantages
and limitations of both testing frameworks, we've advocated for the feasibility of migration and delved into key design
considerations before initiating the transition.

Following this, we've discussed the initial steps of the migration process and prioritized which tests to migrate first.
Recognizing the significance of the migration process itself, we've devoted attention to exploring various implementations
and identifying plugins that could enrich the Cypress experience.

Welcome to the world of smooth testing!

Cypress not only competes effectively with established frameworks like Selenium but also introduces a host of enhancements.
These enhancements encompass easier setup and maintenance, along with significant quality-of-life improvements, as detailed
in the blog post. The minimized flakiness and simplified programming experience have positioned Cypress as a premier Test

> Automation framework, embraced by numerous users, including yourself.

Reviewer: ```suggestion
Automation framework, embraced by numerous users.
```

You can't assume that the reader is already using Cypress. Maybe they are just considering a migration?
